---
title: "repeatToolKit: Network analysis of repetitive elements"
author: "Anthony R. Colombo, Giridharan Ramsingh, Timothy Triche Jr."
output: 
  html_vignette:
  toc: true
  number_sections: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteDepends{arkasData}
  %\VignetteIndexEntry{repeatToolKit: Network analysis for repetitive elements}
  \usepackage[utf8]{inputenc}   
date: "`r format(Sys.time(), '%d %B, %Y')`"
---
          
# Introduction
Arkas software links Kallisto to BioConductor.  Arkas computes Kallisto quantifications accelerating the Kallisto preparatory routines, and executes limma/voom linear models. These linear models can be extended to include transposable elements (TE).  Annotation software TxDbLite identifies differential expression of TE.  
 
 
# Network tools can identify associations between genes and TE
The study of TE and their relationship to genes warrants future studies.  This software, repeatToolKit, provides network tools for analyzing TE associations with genes.

# Creating network tables for genes and TE
The package WGCNA (Horvath,et. al) is wonderful and sophisticated.  It is very useful for identifying preserved clusters of genes, and their associations to external traits such as weight etc.  repeatToolKit adapts WGCNA modules, and identifies hubs of connected genes, in order to measure associations to TEs.  The modules can be explored using RSQLite tables created by repeatToolKit.    

```{r merged_process}
   suppressPackageStartupMessages(library(arkas))
   suppressPackageStartupMessages(library(TxDbLite))
   suppressPackageStartupMessages(library(arkasData))
   suppressPackageStartupMessages(library(repeatToolKit))

  data(NS)
  NS<-annotateFeatures(NS,'transcript')
  NS<-NS[match(sample(rownames(NS),5000),rownames(NS)),]
#  samples<-c("n1","n2","n4","s1","s2","s4")
#  pathBase<-system.file("extdata",package="arkasData")
#  merged <- mergeKallisto(samples, outputPath=pathBase)
  design<-model.matrix(~substring(colnames(NS),1,1))
  rownames(design)<-colnames(NS)
  lnames<-wgcna(NS,whichWGCNA="single",species="Homo.sapiens",intBiotypes=c("Alu","ERV1","ERVK","ERVL","ERV3","hAT","L1","L2","PiggyBac"),tmm.norm=TRUE,how="cpm",design=design,collapseBy="gene_id",selectedPower=6,annotate=TRUE,saveToFile=FALSE)
  rnames<-wgcna(NS,whichWGCNA="single",species="Homo.sapiens",intBiotypes=c("Alu","ERV1","ERVK","ERVL","ERV3","hAT","L1","L2","PiggyBac"),selectedPower=6,tmm.norm=TRUE,how="cpm",design=design,saveToFile=FALSE)  
  ### network analysis scripts
  ## scatter Mod shows the correlation of gene significance and module members, and the MSE.
  wgcna_scatterMod(lnames,biotype=colnames(lnames[["datTraits"]])) 

 ### show module members
  ##identifies highly correlated genes to the module eigenvalue.
  MM1<-showModuleMembers(lnames,biocolor="1")

 ##top hub members of module 1
  ModuleHubGenes(NS,lnames=lnames,module.color="1",connectivity.cutoff=0.85)


```


#Annotations and RPKM/FPKM to Transcripts-Per-Million (TPM) 
Many RNA-Seq experiments derive counts as reads per
kilobase per million, or fragments per kilobase per million. The quantification from RNA-Seq transcript abundances is dependent on transcript length in proportion to relative abundance. The estimated probability of reads generated by a transcript is given by counting the number of reads that align to transcript divided by total number of mapped reads. Where the definition of transcript-per-million, TPM, is given as mean transcript length in kilobases multiplied by RPKM. The mean for transcript dependent lengths gives weighted measure for expression of the lengths across iso-forms. Arkas handles RPKM to TPM conversion by dividing the selected transcript RPKM score by the sum of all RPKM, and multiplying by 1e6 (Li,2009)
TxDbLite is a sister package to arkas which can annotate kallisto-experiment classes.

#Repeat Analysis
 After creating the repeatome, arkas can be used to analyze repetitive elements
 For instance, one can plot repeat element transcripts using (counts/bootstrap MADs) as effect size, or analyze families of Alu, LTR, or Endongenous Retroviruses.  Below we plot a heat map of quantified transcripts from our repeatome for species Homo Sapiens.



#References
Bo, Li, et. al, "RNA-Seq Gene Expression Estimation with Read Mapping Uncertainty."  
Oxford Journals Bioinformatics. Oxford Journal, 9 Dec. 2009. Web. 20 June 2015.   
<http://bioinformatics.oxfordjournals.org/content/26/4/493.full>.

